src/modules/mavlink/mavlink_main.cpp

```diff
@@ -784,40 +784,55 @@ Mavlink::send_packet()
 {
 	int ret = -1;
 
-#if defined(CONFIG_NET) || defined(__PX4_POSIX)
+	// only transmit after we've received messages
+	if (!should_transmit()) {
+		_buf_len = 0;
+	}
 
-	/* Only send packets if there is something in the buffer. */
-	if (_network_buf_len == 0) {
+	// Only send packets if there is something in the buffer
+	if (_buf_len == 0) {
 		pthread_mutex_unlock(&_send_mutex);
 		return 0;
 	}
 
-	if (get_protocol() == UDP) {
+	// send message to UART
+	if (get_protocol() == SERIAL) {
+		// check if there is space in the buffer, let it overflow else
+		if (get_free_tx_buf() < _buf_len) {
+			// not enough space in buffer to send
+			count_txerrbytes(_buf_len);
+			_buf_len = 0;
+			return -1;
+		}
 
-#ifdef CONFIG_NET
+		ret = ::write(_uart_fd, _buf, _buf_len);
+	}
+
+#if defined(CONFIG_NET) || defined(__PX4_POSIX)
+
+	else if (get_protocol() == UDP) {
+
+# if defined(CONFIG_NET)
 
 		if (_src_addr_initialized) {
-#endif
-			ret = sendto(_socket_fd, _network_buf, _network_buf_len, 0,
-				     (struct sockaddr *)&_src_addr, sizeof(_src_addr));
-#ifdef CONFIG_NET
+# endif // CONFIG_NET
+			ret = sendto(_socket_fd, _buf, _buf_len, 0, (struct sockaddr *)&_src_addr, sizeof(_src_addr));
+# if defined(CONFIG_NET)
 		}
 
-#endif
+# endif // CONFIG_NET
 
-		/* resend message via broadcast if no valid connection exists */
+		// resend message via broadcast if no valid connection exists
```

Comment:
The comment is incomplete anyway, better delete it.

Reply:
Removed.